pippy/hf/_SaveModule.py

```diff
@@ -1,11 +1,59 @@
 import torch.distributed as dist
 import pippy
 
+from itertools import chain
+import tempfile
+import logging
+import shutil
 import json
 import os
 
 CKPT_INDEX_JSON_FILENAME = "pytorch_model.bin.index.json"
 
+logger = logging.getLogger(__name__)
+
+
+def _copy_with_metadata(source: str, target: str) -> None:
+    """
+    Copy file with all its permisions and metadata.
+
+    Args:
+        source (str): path to source file
+        target (str): path to copy to
+    """
+    shutil.copy2(source, target)
+    st = os.stat(source)
+    os.chown(target, st.st_uid, st.st_gid)
+
+
+def _atomic_write(file_contents: str, target_file_path: str, mode="w") -> None:
+    """
+    Atomically writes `file_contents` into `target_file_path`.
+
+    Args:
+        file_contents (str): contents to write to file
+        target_file_path (str): path to write to
+        mode (str, optional): mode to write file with. Defaults to "w". Only "w" and "a" are supported.
+    """
+    temp_file = tempfile.NamedTemporaryFile(
+        delete=False, dir=os.path.dirname(target_file_path)
+    )
+    try:
+        if os.path.exists(target_file_path):
+            _copy_with_metadata(target_file_path, temp_file.name)
```

Comment:
To add comment: why do we need a copy first?
nit: what would be the impact on performance?

Reply:
I think in our case, the target_file_path should never exist, as we will be writing the index file during saving. It means that we only need to support both file creation. What do you think?

Reply:
it was just a sanity check in case the user was overwrite a ckpt, but I agree, we can do away with that.

Reply:
Makes sense. Sounds good. Then, let's keep it for now. We can revisit if necessary.